src/python/pants/core/goals/test.py

```diff
@@ -204,6 +218,136 @@ class TestFieldSet(FieldSet, metaclass=ABCMeta):
+            and returns a `Partitions` instance. E.g.
+                @rule
+                async def partition(
+                    request: SuperTesterRequest.PartitionRequest[SuperTesterFieldSet],
+                    subsystem: SuperTesterSubsystem,
+                ) -> Partitions[SuperTesterFieldSet]:
+                    if subsystem.skip:
+                        return Partitions()
+
+                    # One possible implementation
+                    return Partitions.partition_per_input(request.field_sets)
+
+        2. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestResult` instance. E.g.
+                @rule
+                async def super_test(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestResult:
+                    ...
+
+        3. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugRequest:
+                    ...
+
+        4. A rule which takes an instance of your request type's `SubPartition` class property, and
+            returns a `TestDebugAdapterRequest` instance. E.g.
+                @rule
+                async def setup_super_test_debug(
+                    request: SuperTesterRequest.SubPartition[SuperTesterFieldSet],
+                ) -> TestDebugAdapterRequest:
+                    ...
+
+    NOTE: If your tester doesn't support running in debug mode / running with the debug adapter, you
+        can have the `@rule`s in 3 and 4 raise `NotImplementedError`s.
+
+    Lastly, register the rules which tell Pants about your plugin.
+    E.g.
+        def rules():
+            return [
+                *collect_rules(),
+                *SuperTesterRequest.registration_rules(),
+            ]
+    """
+
+    field_set_type: ClassVar[type[TestFieldSet]]
+
+    # Prevent this class from being detected by pytest as a test class.
+    __test__ = False
+
+    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class PartitionRequest(Generic[_FieldSetT]):
+        """Returns a unique `PartitionRequest` type per calling type.
+
+        This serves us 2 purposes:
+            1. `TestRequest.PartitionRequest` is the unique type used as a union base for plugin registration.
+            2. `<Plugin Defined Subclass>.PartitionRequest` is the unique type used as the union member.
+        """
+
+        field_sets: tuple[_FieldSetT, ...]
+
+    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
+    @dataclass(frozen=True)
+    @runtime_ignore_subscripts
+    class SubPartition(Generic[_FieldSetT]):
```

Comment:
Terminology-wise, I'm not clear on the distinction between a Partition and a SubPartition. Can you elaborate in the comments here?

Reply:
That's stolen from the Lint changes to some degree (which doesn't have a `Partition` object, but might if we align).

The reason I named it `SubPartition` over there was because I wanted to convey to the plugin author that the partition they gave us can (and likely will) be sliced into sub-partitions. Type-wise they're identical.

Reply:
It doesn't help I keep chiming in here and in Slack with the gospel of re-use between this and fmt/lint.